File: 2/2220/MinimumBitFlipsToConvertNumber.java
/*
https://leetcode.com/problems/minimum-bit-flips-to-convert-number/description/?envType=daily-question&envId=2024-09-11
*/

class Solution {

    public static String convertToBinary(int number){
        StringBuilder numberToString=null;
        if(number==1||number==0){
            numberToString=new StringBuilder(String.valueOf(number));
        } else{
            numberToString=new StringBuilder("");
            while(number!=1){
                int quotient=number/2, remainder=number%2;
                numberToString.append(String.valueOf(remainder));
                number=quotient;
            }
            numberToString.append("1");
        }
        while(numberToString.length()!=32){
            numberToString.append("0");
        }
        return String.valueOf(numberToString.reverse());
    }

    public int minBitFlips(int start, int goal) {
        String startInBinaryRepresentation=convertToBinary(start);
        String goalInBinaryRepresentation=convertToBinary(goal);
        int resultantMininumBitFlips=0, length=startInBinaryRepresentation.length(), beginIndex=0;
        while(true){
            if(beginIndex>=length) break;
            if(startInBinaryRepresentation.charAt(beginIndex)!=goalInBinaryRepresentation.charAt(beginIndex)){
                resultantMininumBitFlips++;
            }
            beginIndex++;
        }
        return resultantMininumBitFlips;
    }
}

File: 2/2000s/22s/2220/MinimumBitFlipsToConvertNumber.cpp
/*
https://leetcode.com/problems/minimum-bit-flips-to-convert-number/description/?envType=daily-question&envId=2024-09-11
*/

class Solution {
public:

    string toBinary(int number){
        if(number==1||number==0){
            string res=to_string(number);
            while(res.size()!=32){
                res+="0";
            }
            reverse(res.begin(), res.end());
            return res;
        }
        string res="";
        while(number!=1){
            int quo=number/2, rem=number%2;
            res+=to_string(rem);
            number=quo;
        }
        res+="1";
        while(res.size()!=32){
            res+="0";
        }
        reverse(res.begin(), res.end());
        return res;
    }

    int minBitFlips(int start, int goal) {
        string binaryStart=toBinary(start), binaryGoal=toBinary(goal);
        int res=0, sz=binaryStart.size(), idx=0;
        while(true){
            if(idx>=sz) break;
            if(binaryStart[idx]!=binaryGoal[idx]) res++;
            idx++;
        }
        return res;
    }
};